### ISSUE TITLE
Classes and closure semantics

### ISSUE THREAD

**Issue by masak** [masak/007#209]
We're not quite there yet, but just filing this because I got to think about it:

    for [1, 2] -> n {
        my n2 = n;
        class C {
            foo() {
                say(n2);
            }
        }

        (new C {}).foo();    # expecting "1\n" the first iteration and "2\n" the second
    }

In other words, the `outer` of method `foo` should be properly refreshed, just like with ordinary functions. Something tells me this won't happen on its own. It didn't with functions.

**Comment by masak** [comment]
Just going to point out here one consequence of us doing this. (And possibly the biggest reason Perl 6 has gone with "the other" semantics.)

    sub foo() {
        class C {}
        return new C {};
    }
    say(foo().type == foo().type);    # False

At least I *think* that's what it means. It's remotely possible that we might get to have our cake and eat it here.

But even if we can't, I think on balance I still prefer to have the class closure semantics and sacrifice everywhere type equality, rather than the other way around.

**Comment by masak** [comment]
...the problematic thing about `foo().type == foo().type` being `False` is not so much that someone might expect that to be true in their program and be disappointed it isn't, but that it seems to imply that a type is generated for the class `C` twice at runtime.

And _that_ is a bridge I don't think I want to cross. I'm still (tentatively) all for reifying our types at runtime, but I don't want to go about *creating* them at runtime. At least not by default. A textual construct like `class C {}` should _definitely_ have done its job with creating its underlying `.type` already at compile time.

So `foo().type == foo().type` _has_ to come out `True`. If that means that class closure semantics goes out the window, then so be it. (But I'm still not sure it does. It still seems to me like the class closure could refresh at each surrounding block entry, while the class still retains its `.type`.)

**Comment by masak** [comment]
Getting closer to what's actually impossible here: the methods close over the surrounding scope, so they'd have to refresh on surrounding block entry. (So would initializer expressions.) Where do these live? In the `Type` object. Combine that with things closing over the type object, and... I think that forces us to have several type objects.

Node does it that way:

```
> function foo(x) { class C { bar() { return x } }; return new C() }
> foo(1).bar()
1
> foo(2).bar()
2
```

This feels a bit "too dynamic" for 007. Part of the charm of classes is that we sort of shunt some of the complexity in the static/compile-time direction. I don't want to sell that advantage off too cheaply.

But we're in a situation where we have to choose between these:

1. Not allow classes to close over external identifiers at all. (Unsatisfactory, and unacceptable, I think.)
2. Allow them to close over external identifiers, but only from the static block. (Surprising.)
3. Create the type objects at runtime.

**Comment by masak** [comment]
Even if we went with the most restrictive, boringest solution of having classes and their type objects be _very_ static in 007, we'll probably need some kind of "fixup" solution, just so

    perl6 -e'my $x; class C { method foo { say $x } }; $x = "OH HAI"; C.new.foo'
    OH HAI

will resolve the `$x` lookup in the method `foo` to the runtime scope where `$x` has the value `"OH HAI"`, even though as `C` is defined (at `BEGIN` time), this scope does not exist.

**Comment by masak** [comment]
Python sides with Node and makes classes a dynamic thing:

    $ python3
    Python 3.5.3 (default, Sep 27 2018, 17:25:39)
    [GCC 6.3.0 20170516] on Linux
    Type "help", "copyright", "credits" or "license" for more information
    >>> def foo(x):
    ...     class C:
    ...         def bar(self):
    ...             return x
    ...     return C()
    ...
    >>> foo(1).bar()
    1
    >>> foo(2).bar()
    2

Even Perl 6 does this, so I guess this is more to do with the fixup thing than with anything else:

    $ perl6
    You may want to `zef install Readline` or `zef install Linenoise` or use rlwrap for a line editor
    
    To exit type `exit` or `^D`
    > sub foo($x) {
    >   class C {
    >     method bar() { $x }
    >   }
    > }
    &foo
    > foo(1).bar()
    1
    > foo(2).bar()
    2

but

    > my $f1 = foo(1)
    C.new
    > my $f2 = foo(2)
    C.new
    > $f1.bar()
    2
    > $f2.bar()
    2

So I think a fair description is that in Perl 6, classes nested in a function do _not_ have full closure semantics... but they get pretty far simply by fixing up the OUTER of a nested class.

In contrast, Python's class definitions are evaluated at runtime and they have full closure semantics:

    >>> f1 = foo(1)
    >>> f2 = foo(2)
    >>> f1.bar()
    1
    >>> f2.bar()
    2

**Comment by masak** [comment]
> Getting closer to what's actually impossible here: the methods close over the surrounding scope, so they'd have to refresh on surrounding block entry. (So would initializer expressions.) Where do these live? In the `Type` object. Combine that with things closing over the type object, and... I think that forces us to have several type objects.

No. The future me is not convinced by this argument.

It seems to me the following two things could co-exist:

* Type objects, which are declared statically (even if they are in a nested scope and lexically closes over some names); here, by "statically", I mean something in the vicinity of parse-time, or maybe the MOP has its own phase sometime later &mdash; in particular, for a `class C` (even one in a `for` loop), there could be exactly one Type object.
* Class bodies, acting like closures, in that the `OUTER` pointer of the block of `class C` (and all the methods and initializers in it) is always correctly pointing to its unique dynamically entered containing frame.

In other words, type objects are static and unique, but class bodies are dynamic and refreshing.

**Comment by vendethiel** [comment]
Of note, the `$n` is shared between instances:

```raku
my @c;
for 1..2 -> $n {
	my class C {
		method foo() { $n }
	}
        @c.push: C;
}
say @c[0].foo; # 2
```

**Comment by masak** [comment]
Oh right; as was pointed out before. So Raku doesn't really exhibit the co-existence of unique type objects and class closures. D'oh.

And maybe it isn't actually possible? When you ask "what is `C`?", the answer kind of has to be _either_ the unique type object _or_ the just-created class closure. Raku makes it the former; it's the "fixup mechanism" which then makes the answer in that code `2`, not `1`.
